modules/porous_flow/src/bcs/PorousFlowSink.C

```diff
@@ -0,0 +1,179 @@
+  params.addRequiredParam<unsigned int>("fluid_phase", "The index corresponding to the fluid phase for this BC");
+  params.addParam<unsigned int>("mass_fraction_component", "The index corresponding to a fluid component.  If supplied, the flux will be multiplied by the nodal mass fraction for the component");
+  params.addParam<bool>("use_mobility", false, "If true, then fluxes are multiplied by (density*permeability_nn/viscosity), where the '_nn' indicates the component normal to the boundary.  In this case bare_flux is measured in Pa.s^-1.  This can be used in conjunction with use_relperm.");
+  params.addParam<bool>("use_relperm", false, "If true, then fluxes are multiplied by relative permeability.  This can be used in conjunction with use_mobility");
+  params.addParam<FunctionName>("flux_function", 1.0, "The flux.  The flux is OUT of the medium: hence positive values of this function means this BC will act as a SINK, while negative values indicate this flux will be a SOURCE.  The functional form is useful for spatially or temporally varying sinks. If use_mobility=false then this function is measured in kg.m^-2.s^-1.  If use_mobility=true then it is measured in Pa.s^-1.");
+  return params;
+}
+
+PorousFlowSink::PorousFlowSink(const InputParameters & parameters) :
+    IntegratedBC(parameters),
+    _dictator(getUserObject<PorousFlowDictator>("PorousFlowDictator")),
+    _ph(getParam<unsigned int>("fluid_phase")),
+    _use_mass_fraction(isParamValid("mass_fraction_component")),
+    _sp(_use_mass_fraction ? getParam<unsigned int>("mass_fraction_component") : 0),
+    _use_mobility(getParam<bool>("use_mobility")),
+    _use_relperm(getParam<bool>("use_relperm")),
+    _m_func(getFunction("flux_function")),
+    _permeability(getMaterialProperty<RealTensorValue>("PorousFlow_permeability")),
+    _dpermeability_dvar(getMaterialProperty<std::vector<RealTensorValue> >("dPorousFlow_permeability_dvar")),
+    _fluid_density_node(getMaterialProperty<std::vector<Real> >("PorousFlow_fluid_phase_density")),
+    _dfluid_density_node_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorousFlow_fluid_phase_density_dvar")),
+    _fluid_viscosity(getMaterialProperty<std::vector<Real> >("PorousFlow_viscosity")),
+    _dfluid_viscosity_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorousFlow_viscosity_dvar")),
+    _relative_permeability(getMaterialProperty<std::vector<Real> >("PorousFlow_relative_permeability")),
+    _drelative_permeability_dvar(getMaterialProperty<std::vector<std::vector<Real> > >("dPorousFlow_relative_permeability_dvar")),
+    _mass_fractions(getMaterialProperty<std::vector<std::vector<Real> > >("PorousFlow_mass_frac")),
+    _dmass_fractions_dvar(getMaterialProperty<std::vector<std::vector<std::vector<Real> > > >("dPorousFlow_mass_frac_dvar")),
+    _node_number(getMaterialProperty<unsigned int>("PorousFlow_node_number")),
+    _qp_map(0)
```

Comment:
This can be a little bit confusing. The default c-tor should do the right thing, i.e. you would not list `_qp_map` here at all...